File: MiniDOMFinanziaria/Finanziaria/Pratiche/StatiPratica/VincoliStatoPratica/CStatoPratRulesCollection.cs
using System;

namespace minidom
{
    public partial class Finanziaria
    {


        /// <summary>
    /// Insieme di oggetti CStatoPratRule definiti su uno stato
    /// </summary>
    /// <remarks></remarks>
        [Serializable]
        public class CStatoPratRulesCollection : CCollection<CStatoPratRule>
        {
            [NonSerialized]
            private CStatoPratica m_Parent; // [CStatoPratica] Oggetto a cui appartiene la collezione

            public CStatoPratRulesCollection()
            {
                m_Parent = null;
            }

            public CStatoPratRulesCollection(CStatoPratica statoIniziale) : this()
            {
                Load(statoIniziale);
            }

            public CStatoPratica Parent
            {
                get
                {
                    return m_Parent;
                }
            }

            protected override void OnInsert(int index, object value)
            {
                if (m_Parent is object)
                    ((CStatoPratRule)value).SetSource(m_Parent);
                base.OnInsert(index, value);
            }

            protected override void OnSet(int index, object oldValue, object newValue)
            {
                if (m_Parent is object)
                    ((CStatoPratRule)newValue).SetSource(m_Parent);
                base.OnSet(index, oldValue, newValue);
            }

            internal void Load(CStatoPratica parent)
            {
                lock (this)
                {
                    if (parent is null)
                        throw new ArgumentNullException("parent");
                    Clear();
                    m_Parent = parent;
                    if (DBUtils.GetID(parent) == 0)
                        return;

                    // Dim cursor As New CStatoPratRuleCursor
                    // cursor.IDSource.Value = DBUtils.GetID(parent, 0)
                    // cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                    // cursor.IgnoreRights = True
                    // While Not cursor.EOF
                    // MyBase.Add(cursor.Item)
                    // cursor.MoveNext()
                    // End While
                    // cursor.Reset()
                    // cursor = Nothing
                    lock (StatiPratRules)
                    {
                        foreach (CStatoPratRule rule in StatiPratRules.LoadAll())
                        {
                            if (rule.Stato == ObjectStatus.OBJECT_VALID && rule.IDSource == DBUtils.GetID(parent))
                            {
                                Add(rule);
                            }
                        }
                    }

                    Sort();
                }
            }

            protected internal virtual void SetOwner(CStatoPratica owner)
            {
                m_Parent = owner;
                if (owner is null)
                    return;
                foreach (CStatoPratRule rule in this)
                    rule.SetSource(owner);
            }
        }
    }
}

File: MiniDOMFinanziaria/Finanziaria/Pratiche/StatiPratica/VincoliStatoPratica/CStatoPratRulesCollection.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Anagrafica



Partial Public Class Finanziaria


    ''' <summary>
    ''' Insieme di oggetti CStatoPratRule definiti su uno stato
    ''' </summary>
    ''' <remarks></remarks>
    <Serializable> _
    Public Class CStatoPratRulesCollection
        Inherits CCollection(Of CStatoPratRule)

        <NonSerialized> _
        Private m_Parent As CStatoPratica '[CStatoPratica] Oggetto a cui appartiene la collezione

        Public Sub New()
            Me.m_Parent = Nothing
        End Sub

        Public Sub New(ByVal statoIniziale As CStatoPratica)
            Me.New()
            Me.Load(statoIniziale)
        End Sub

        Public ReadOnly Property Parent As CStatoPratica
            Get
                Return Me.m_Parent
            End Get
        End Property

        Protected Overrides Sub OnInsert(index As Integer, value As Object)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(value, CStatoPratRule).SetSource(Me.m_Parent)
            MyBase.OnInsert(index, value)
        End Sub

        Protected Overrides Sub OnSet(index As Integer, oldValue As Object, newValue As Object)
            If (Me.m_Parent IsNot Nothing) Then DirectCast(newValue, CStatoPratRule).SetSource(Me.m_Parent)
            MyBase.OnSet(index, oldValue, newValue)
        End Sub

        Friend Sub Load(ByVal parent As CStatoPratica)
            SyncLock Me
                If (parent Is Nothing) Then Throw New ArgumentNullException("parent")
                MyBase.Clear()
                Me.m_Parent = parent
                If (GetID(parent) = 0) Then Return

                'Dim cursor As New CStatoPratRuleCursor
                'cursor.IDSource.Value = Databases.GetID(parent, 0)
                'cursor.Stato.Value = ObjectStatus.OBJECT_VALID
                'cursor.IgnoreRights = True
                'While Not cursor.EOF
                '    MyBase.Add(cursor.Item)
                '    cursor.MoveNext()
                'End While
                'cursor.Reset()
                'cursor = Nothing
                SyncLock Finanziaria.StatiPratRules
                    For Each rule As CStatoPratRule In Finanziaria.StatiPratRules.LoadAll
                        If rule.Stato = ObjectStatus.OBJECT_VALID AndAlso rule.IDSource = GetID(parent) Then
                            Me.Add(rule)
                        End If
                    Next
                End SyncLock
                Me.Sort()
            End SyncLock
        End Sub

        Protected Friend Overridable Sub SetOwner(ByVal owner As CStatoPratica)
            Me.m_Parent = owner
            If (owner Is Nothing) Then Exit Sub
            For Each rule As CStatoPratRule In Me
                rule.SetSource(owner)
            Next
        End Sub

    End Class



End Class
